From: "Igor B. Poretsky" <poretsky@rakurs.com>
Date: Thu, 24 Nov 2022 15:27:36 +0300
Subject: MPlayer frontend enhancements

* More comprehensive feedback when marking clips.

* Conventional file path input.

* MPlayer responses parsing stability fix.

* Mediatrack length getting function.

* Media track splitting functionality.

Clip extraction functionality improved:

- Sox error messages are preserved in case of failure.
- User is asked about output file overwriting.
---
 lisp/emacspeak-m-player.el | 170 ++++++++++++++++++++++++++++++++++-----------
 1 file changed, 130 insertions(+), 40 deletions(-)

diff --git a/lisp/emacspeak-m-player.el b/lisp/emacspeak-m-player.el
index f47ced5..cea6e7f 100644
--- a/lisp/emacspeak-m-player.el
+++ b/lisp/emacspeak-m-player.el
@@ -252,6 +252,13 @@ on a specific directory."
   (regexp-opt '("mp3" "audio" "music"))
   "Pattern matching locations where we store media.")
 
+(defcustom emacspeak-m-player-coding-system nil
+  "Media player output coding system.
+It is used for tags decoding."
+  :type '(coding-system :size 0)
+  :group 'emacspeak-m-player)
+
+
 ;;;###autoload
 (defun emacspeak-multimedia  ()
   "Start or control Emacspeak multimedia player.
@@ -365,10 +372,6 @@ etc to be ignored when guessing directory.")
 (defun emacspeak-media-read-resource ()
   "Read resource from minibuffer with contextual smarts."
   (let ((completion-ignore-case t)
-        (read-file-name-function
-         (if (eq major-mode 'locate-mode)
-             #'read-file-name-default
-           #'ido-read-file-name))
         (read-file-name-completion-ignore-case t)
         (default-filename
           (when (or (eq major-mode 'dired-mode) (eq major-mode 'locate-mode))
@@ -405,6 +408,7 @@ etc to be ignored when guessing directory.")
   "Filter function that captures metadata.
 Also cleanup ANSI escape sequences."
   (cl-declare (special emacspeak-m-player-cue-info
+                       ansi-color-drop-regexp
                        ansi-color-control-seq-regexp))
   (when (process-live-p process)
     (with-current-buffer (process-buffer process)
@@ -418,11 +422,14 @@ Also cleanup ANSI escape sequences."
           (emacspeak-auditory-icon 'progress)
           (emacspeak-m-player-stream-info)))
       (goto-char (process-mark process))
-      (let ((start (point)))
+      (let ((start (point))
+            (ignore-seq-regexp
+             (or (and (boundp 'ansi-color-control-seq-regexp) ansi-color-control-seq-regexp)
+                 (and (boundp 'ansi-color-drop-regexp) ansi-color-drop-regexp))))
         (insert output)
         (save-mark-and-excursion
           (goto-char start)
-          (while (re-search-forward ansi-color-control-seq-regexp  (point-max) 'no-error)
+          (while (and ignore-seq-regexp (re-search-forward ignore-seq-regexp  (point-max) 'no-error))
             (delete-region (match-beginning 0) (match-end 0))))))))
 
 (defun emacspeak-m-player-amark-save ()
@@ -499,6 +506,8 @@ The player is placed in a buffer in emacspeak-m-player-mode."
             (apply
              #'start-process "MPLayer" buffer
              emacspeak-m-player-program options))
+      (set-process-coding-system emacspeak-m-player-process
+                                 emacspeak-m-player-coding-system)
       (set-process-filter  emacspeak-m-player-process
                            #'emacspeak-m-player-process-filter)
       (when (and
@@ -621,32 +630,52 @@ necessary."
 ;;}}}
 ;;{{{ commands
 
+(defconst emacspeak-m-player-info-order
+  '("TIME_POSITION"
+    "PERCENT_POSITION"
+    "LENGTH"
+    "FILENAME")
+  "MPlayer info items sorting order.")
+
+(defun emacspeak-m-player-info-priority (item)
+  "Returns sorting priority for specified item."
+  (let ((order emacspeak-m-player-info-order)
+        (result 0))
+    (while (and order (not (string-match (car order) (car item))))
+      (setq result (1+ result)
+            order (cdr order)))
+    result))
+
+(defun emacspeak-m-player-info-item-precedes-p (first second)
+  "Indicates that the first item should precede the second one."
+  (<= (emacspeak-m-player-info-priority first)
+      (emacspeak-m-player-info-priority second)))
+
 (defun emacspeak-m-player-get-position ()
   "Return list suitable to use as an amark. --- see emacspeak-amark.el.
 Return value is of the form (position filename length)."
-  (cl-declare (special emacspeak-m-player-process))
-  (with-current-buffer (process-buffer emacspeak-m-player-process)
-;;; dispatch command twice to avoid flakiness in mplayer
-    (emacspeak-m-player-dispatch "get_time_pos\nget_file_name\nget_time_length\n")
-    (let* ((output (emacspeak-m-player-dispatch "get_time_pos\nget_file_name\nget_time_length\n") )
-           (lines (split-string output "\n" 'omit-nulls))
-           (fields
-            (cl-loop
-             for l in lines
-             collect (cl-second (split-string l "=")))))
-      (list
-       (format "%s" (cl-first fields))  ; position
-       (if (cl-second fields)
-           (substring (cl-second  fields) 1 -1)
-         "")
-       (format "%s" (cl-third fields))))))
+  (let* ((output (emacspeak-m-player-dispatch "get_time_pos\nget_file_name\n"))
+         (lines (split-string output "\n" 'omit-nulls))
+         (fields
+          (sort
+           (cl-loop
+            for l in lines
+            collect (split-string l "="))
+           'emacspeak-m-player-info-item-precedes-p)))
+    (list
+     (format "%s" (cl-second (cl-first fields)))     ; position
+     (if (cl-second fields)
+         (substring (cl-second (cl-second fields)) 1 -1)
+       ""))))
 
 (defun emacspeak-m-player-current-filename ()
   "Return filename of currently playing track."
-  (cl-second
-   (split-string
-    (emacspeak-m-player-dispatch "get_file_name\n")
-    "=")))
+  (substring
+   (cl-second
+    (split-string
+     (emacspeak-m-player-dispatch "get_file_name\n")
+     "="))
+   1 -1))
 
 (defun emacspeak-m-player-scale-speed (factor)
   "Scale speed by specified factor."
@@ -917,9 +946,19 @@ Interactive prefix arg toggles automatic cueing of ICY info updates."
 
 
 (defun emacspeak-m-player-get-length ()
+  "Get length of track in seconds."
+  (cl-second
+   (split-string
+    (emacspeak-m-player-dispatch "get_time_length")
+    "=")))
+
+;;;###autoload
+(defun emacspeak-m-player-display-length ()
   "Display length of track in seconds."
   (interactive)
-  (dtk-speak-and-echo (emacspeak-m-player-dispatch "get_time_length")))
+  (dtk-speak-and-echo
+   (format "Total track length is %s seconds"
+           (emacspeak-m-player-get-length))))
 
 (defconst emacspeak-m-player-display-cmd
   "get_time_pos\nget_percent_pos\nget_time_length\nget_file_name\n"
@@ -934,9 +973,11 @@ Interactive prefix arg toggles automatic cueing of ICY info updates."
     (when result
       (setq result (replace-regexp-in-string  "^ans_" "" result))
       (setq fields
-            (mapcar
-             #'(lambda (s) (split-string s "="))
-             (split-string  result "\n"))))
+            (sort
+             (mapcar
+              #'(lambda (s) (split-string s "="))
+              (split-string  result "\n"))
+             'emacspeak-m-player-info-item-precedes-p)))
     (cond
      (fields                       ; speak them after audio formatting
       (cl-loop
@@ -1270,6 +1311,7 @@ flat classical club dance full-bass full-bass-and-treble
     ("Q" emacspeak-m-player-quit)
     ("R" emacspeak-m-player-edit-reverb)
     ("S" emacspeak-m-player-amark-save)
+    ("X" emacspeak-m-player-export-track-list)
     ("x" emacspeak-m-player-pan)
     ("w" emacspeak-m-player-write-clip)
     ("SPC" emacspeak-m-player-pause)
@@ -1420,7 +1462,46 @@ As the default, use current position."
   "Return index of name in file-list."
   (cl-position (expand-file-name name) file-list :test #'string=))
 
+(defun emacspeak-m-player-amark-precedes-p (first second)
+  "Indicates that the first amark precedes the second one."
+  (<= (string-to-number (emacspeak-amark-position first))
+      (string-to-number (emacspeak-amark-position second))))
 
+(defun emacspeak-m-player-export-track-list ()
+  "Save track split data in Audacity labels file using amarks as split points."
+  (interactive)
+  (cl-declare (special buffer-undo-list emacspeak-amark-list))
+  (let* ((file (emacspeak-m-player-current-filename))
+         (tracklist (expand-file-name (format "%s.tracks" file)))
+         (amarks
+          (sort
+           (cl-loop for amark in emacspeak-amark-list
+                    when (string-equal (emacspeak-amark-path amark) file)
+                    collect amark)
+           'emacspeak-m-player-amark-precedes-p)))
+    (if (and amarks
+             (or (not (file-exists-p tracklist))
+                 (y-or-n-p (format "File `%s' already exists. Overwrite? " tracklist))))
+        (let ((buf (find-file-noselect tracklist)))
+          (with-current-buffer buf
+            (setq buffer-undo-list t)
+            (erase-buffer)
+            (while amarks
+              (insert
+               (format "%s\t%s\t%s\n"
+                       (emacspeak-amark-position (car amarks))
+                       (if (cdr amarks)
+                           (emacspeak-amark-position (cadr amarks))
+                         (emacspeak-m-player-get-length))
+                       (emacspeak-amark-name (car amarks))))
+              (setq amarks (cdr amarks)))
+            (goto-char (point-min))
+            (save-buffer)
+            (kill-buffer buf)
+            (emacspeak-auditory-icon 'save-object)))
+      (error "No amarks to export."))))
+
+;;;###autoload
 (defun emacspeak-m-player-amark-jump ()
   "Jump to specified AMark."
   (interactive)
@@ -1846,7 +1927,7 @@ our pre-defined filters if appropriate."
   (setq emacspeak-m-player-clip-start
         (read (cl-first (emacspeak-m-player-get-position))))
   (when  (called-interactively-p 'interactive)
-    (message "mark set at %s" emacspeak-m-player-clip-start)
+    (message "Clip start set at %s" emacspeak-m-player-clip-start)
     (emacspeak-auditory-icon 'mark-object)))
 
 (defun emacspeak-m-player-set-clip-end    ()
@@ -1856,11 +1937,11 @@ our pre-defined filters if appropriate."
   (setq emacspeak-m-player-clip-end
         (read (cl-first (emacspeak-m-player-get-position))))
   (when  (called-interactively-p 'interactive)
-    (message "mark set at %s" emacspeak-m-player-clip-end)
+    (message "Clip end set at %s" emacspeak-m-player-clip-end)
     (emacspeak-auditory-icon 'mark-object)))
 
 (defun emacspeak-m-player-write-clip ()
-  "Invoke mp3splt to clip selected range in current file."
+  "Invoke sox to clip selected range in current file."
   (interactive)
   (cl-declare (special emacspeak-sox
                        emacspeak-m-player-clip-end emacspeak-m-player-clip-start))
@@ -1868,13 +1949,22 @@ our pre-defined filters if appropriate."
   (cl-assert (eq major-mode 'emacspeak-m-player-mode) nil "Not in an MPlayer buffer.")
   (cl-assert (numberp emacspeak-m-player-clip-start) nil "Set start of clip with M-[")
   (cl-assert (numberp emacspeak-m-player-clip-end) nil "Set end of clip with M-]")
-  (let ((file (cl-second (emacspeak-m-player-get-position))))
-    (shell-command
-     (format "%s '%s' 'clip-%s'  trim %s %s"
-             emacspeak-sox file file 
-             emacspeak-m-player-clip-start
-             (- emacspeak-m-player-clip-end emacspeak-m-player-clip-start)))
-    (message "Wrote clip to clip-%s" file)))
+  (let* ((file (emacspeak-m-player-current-filename))
+         (clip (format "clip-%s" file))
+         (clip-start emacspeak-m-player-clip-start)
+         (clip-end emacspeak-m-player-clip-end))
+    (when (or (not (file-exists-p clip))
+              (y-or-n-p (format "File `%s' already exists. Overwrite? " clip)))
+      (with-temp-buffer
+        (unless (zerop (call-process emacspeak-sox nil t nil
+                                     file
+                                     clip
+                                     "trim"
+                                     (number-to-string clip-start)
+                                     (number-to-string (- clip-end clip-start))))
+          (error "Cannot write clip: %s" (buffer-string))))
+      (emacspeak-auditory-icon 'save-object)
+      (message "Wrote clip to %s" clip))))
 
 ;;}}}
 (provide 'emacspeak-m-player)
